Set up Flutter-Django API communication

wozlsla opened the pull request:
## Frontend (Flutter)

```
.
├── common
│   ├── utils
│   │   ├── api.dart
│   │   └── utils.dart
│   └── widgets
│       └── bottom_nav_bar.dart
├── main.dart
├── providers
│   └── tab_provider.dart
├── route
│   ├── router.dart
│   └── router_const.dart
└── views
    ├── entrypoint
    │   └── entrypoint.dart
    ├── home
    │   └── home_screen.dart
    └── note
        ├── note_list_screen.dart
        ├── note_screen.dart
        └── temp_screen.dart
```
- EntryPoint, BottomNavBar(custom), StatefulShellRoute 
  - 다른 스크린들과 분리 목적
  - router, provider 추가

> [!NOTE]  
> App과 Web의 routing 문제로 고민
> Web: (get)/note (post)note/create -> App 에서는 반대로 동작하길 원함. 고려 후 수정 예정

<br/>

## Backend
아직 서버에는 배포 X
<br/>

### Apps
Users, Notes
- Custom user model 생성, 로그인 나중에 구현할거라서 api는 생성 X
- Notes 는 flutter 연동 확인용 (완료)
<br/>

### API

| Endpoint  | 상태    | 설명                                             |
| --------- | ------- | ------------------------------------------------ |
| `/users/` | 개발 중 | 사용자 인증/정보 관련 API 예정                   |
| `/notes/` | 테스트  | Flutter에서 노트 등록/조회 기능 연동 테스트 완료 |

<br/>

```http
GET /api/v1/notes/
POST /api/v1/notes/
```

wozlsla committed:
feat(backend): init django project with env-based SECRET_KEY management

wozlsla committed:
feat(back): add custom user model

wozlsla committed:
feat(back): add Note model and API for Flutter integration test

TEST
- Note 모델 생성
- DRF ViewSet, Serializer 구현
- URL 라우팅 /api/vi/notes/

wozlsla committed:
feat(front): connect to Django API and implement note create & fetch UI

wozlsla committed:
feat(front): add entrypoint and StatefulShellRoute with router/provier

ggomdong commented on backend/pyproject.toml:
```diff
@@ -0,0 +1,20 @@
+description = ""
+authors = [
+    {name = "jimin",email = "jiminseo0307@gmail.com"}
+]
```
보안상 제거가 필요해보입니다. 당장 저만해도 이걸로 많은 정보를 얻을 수 있었네요. "짐니"의 어원도...... ㅎㅎ;

wozlsla replied:
앗 들켜버렸습니다... 이스터 에그..가 아니고, 이건 앞으로의 교훈으로 삼겠습니다!!

ggomdong commented on frontend/lib/route/router.dart:
```diff
@@ -0,0 +1,58 @@
+import 'package:frontend/views/note/temp_screen.dart';
+import 'package:go_router/go_router.dart';
+
+final GlobalKey<NavigatorState> _rootNavigatorKey = GlobalKey<NavigatorState>();
```
_rootNavigatorKey는 현재 사용하시나요? 역할이 없어보이는데.. 향후 사용을 위해 두신건지 궁금합니다.

wozlsla replied:
이후에 유용하게 쓰일 거라고 생각해서 미리 추가해두었습니다!! 이전 프로젝트에서 썼던 기억이 희미하지만 남아있습니다..

ggomdong commented on frontend/lib/views/entrypoint/entrypoint.dart:
```diff
@@ -0,0 +1,48 @@
+    final selectedIndex = ref.watch(tabIndexProvider);
+
+    return Scaffold(
+      body: widget.navigationShell,
```
AppBar는 각 화면에서 정의하나요? 현재 각 화면에도 AppBar는 없더라구요.(이건 그냥 질문입니다. ㅎㅎ)

wozlsla replied:
Appbar는 아직 어떻게 보여줄지 못 정했습니다.. 지금 고려중인 컨셉에 일반적인 appbar 형태 보단 좀 바꿔보고 싶어서요. 하지만 아이디어가 잘 안나오네요 ㅠ,

ggomdong commented on frontend/lib/views/note/note_list_screen.dart:
```diff
@@ -0,0 +1,89 @@
+          ),
+        ],
+      );
+    } else if (notes.isEmpty) {
+      body = const Center(child: Text('노트 없음'));
+    } else {
```
결국 2개의 조건은 모두 status= NoteFetchStatus.success 에 해당하는거니까요. 하나로 묶는게 가독성 측면에서 낫지 않을까 싶습니다.
} else if(status= NoteFetchStatus.success) {
   if (notes.isEmpty) {
   } else {
   }
}

* 물론 else if(status= NoteFetchStatus.success) 는 그 외의 조건이 없다면 그냥 else 만 써도 됩니다.

wozlsla replied:
수정했습니다! 그 과정에서 강의/챌린지 중에는 '구현'에 초점을 맞췄었는데, 이제와서보니 그 구조를 더 잘 이해 '해야한다'는 것을 느끼는 중입니다..

ggomdong commented on frontend/lib/views/note/note_list_screen.dart:
```diff
@@ -0,0 +1,89 @@
+    try {
+      final response = await http.get(notesUrl);
+
+      if (response.statusCode == 200) {
+        setState(() {
+          notes = jsonDecode(utf8.decode(response.bodyBytes));
+          status = NoteFetchStatus.success;
+        });
+      } else {
+        setState(() {
+          status = NoteFetchStatus.error;
+        });
+        print('노트 불러오기 실패: ${response.statusCode}');
+      }
+    } catch (e) {
+      setState(() {
+        status = NoteFetchStatus.error;
+      });
```
setState()는 마지막에 한번만 호출해도 되지 않을까요?

wozlsla replied:
많이 배웠습니다.. 두 방향 다 같은 횟수로 호출된다고 생각해서 별 뜻을 두지 않았었는데, 꼼동님 조언대로 리펙토링 후 훨씬 개선된 것을 느꼈습니다!!

ggomdong commented on frontend/lib/views/note/note_screen.dart:
```diff
@@ -0,0 +1,68 @@
+}
+
+class _NoteScreenState extends State<NoteScreen> {
+  final titleController = TextEditingController();
+  final contentsController = TextEditingController();
```
dispose() 가 필요합니다.

ggomdong commented on frontend/lib/views/note/note_screen.dart:
```diff
@@ -0,0 +1,63 @@
+  final contentsController = TextEditingController();
+
+  Future<void> submitNote() async {
+    final response = await http.post(
```
http에 대한 예외처리를 위해 try~catch 가 필요해 보입니다.

wozlsla replied:
엄청 중요한 것을 빼먹었네요.. 다시는 이런 실수 없도록... 
이 모든 영광을 꼼동님께 돌립니다.

ggomdong commented on frontend/lib/views/note/note_screen.dart:
```diff
@@ -0,0 +1,63 @@
+      notesUrl, // iOS
+      headers: {'Content-Type': 'application/json'},
+      body: jsonEncode({
+        'title': titleController.text,
+        'contents': contentsController.text,
```
입력값에 대한 유효성 검증이 필요합니다.(아무값도 넣지 않았다거나..)

wozlsla replied:
지금은 테스트지만, 이후에 실제로 구성할 때 이 부분 모두 참고하려고 임시지만 전부 추가했습니다! 감사합니다~!!!🤩

ggomdong reviewed (commented):
일단 제겐 너무 어렵고 멋진 코드가 많네요. 수고 많으셨습니다!!!
핵심파트가 아니라서 대략적으로 구현하신 부분도 있으실테지만, 혹시 몰라서 사소한 코멘트들을 좀 달아놨습니다. 중요한게 아니면 무시하셔도 됩니다. ^^

wozlsla committed:
refactor: improve note submission structure and async safety

- 각각 호출되던 setState()를 한 번으로 통합
- status 조건 중복 부분 수정
- http.post() 호출에 try-catch 블록을 추가하여 네트워크 예외 확인
- title, contents 입력값에 대해 .trim() 및 isEmpty 기반의 유효성 검사 추가
- await 이후의 BuildContext 사용 시 mounted 여부를 확인, 위젯이 dispose된 상태에서의 접근 방지
- logging 추가 (사용 고려 중)
